From 8a53b72c7b26fac32ac2856f7cc8583df8a0ee28 Mon Sep 17 00:00:00 2001
From: Zenghui Yu <yuzenghui@huawei.com>
Date: Tue, 13 Dec 2022 22:27:59 +0800
Subject: [PATCH] KVM: arm64: Fix {fp_asimd,sve}_exit_stat manipulation
Patch-mainline: Not yet, from openEuler
References: bsn#22
openEuler-commit: 8a53b72c7b26fac32ac2856f7cc8583df8a0ee28
Modified-by-SEL: No


virt inclusion
category: bugfix
bugzilla: https://gitee.com/openeuler/kernel/issues/I65EGT
CVE: NA

----------------------------------------------------

Currently fp_asimd_exit_stat is accumulated for *both* FP/ASIMD and SVE
traps so that user can not distinguish between these two via debugfs.

Fix the manipulation for both exception classes.

Signed-off-by: Zenghui Yu <yuzenghui@huawei.com>
Reviewed-by: Keqian Zhu <zhukeqian1@huawei.com>
Signed-off-by: Zheng Zengkai <zhengzengkai@huawei.com>
Signed-off-by: Guoqing Jiang <guoqing.jiang@suse.com>
---
 arch/arm64/kvm/hyp/include/hyp/switch.h | 6 +++++-
 1 file changed, 5 insertions(+), 1 deletion(-)

diff --git a/arch/arm64/kvm/hyp/include/hyp/switch.h b/arch/arm64/kvm/hyp/include/hyp/switch.h
index 5d1973b58e3c..f4938db6c71c 100644
--- a/arch/arm64/kvm/hyp/include/hyp/switch.h
+++ b/arch/arm64/kvm/hyp/include/hyp/switch.h
@@ -223,7 +223,11 @@ static inline bool __hyp_handle_fpsimd(struct kvm_vcpu *vcpu)
 	    esr_ec != ESR_ELx_EC_SVE)
 		return false;
 
-	vcpu->stat.fp_asimd_exit_stat++;
+	if (esr_ec == ESR_ELx_EC_FP_ASIMD)
+		vcpu->stat.fp_asimd_exit_stat++;
+	else	/* SVE trap */
+		vcpu->stat.sve_exit_stat++;
+
 	/* Don't handle SVE traps for non-SVE vcpus here: */
 	if (!sve_guest)
 		if (esr_ec != ESR_ELx_EC_FP_ASIMD)
-- 
2.33.0

